docs/design/specs/Ecma-335-Augments.md

```diff
@@ -948,3 +949,12 @@ Conversions from floating-point numbers to integral values truncate the number t
 on the top of the stack is reinterpreted as an unsigned value before the conversion.
 Note that integer values of less than 4 bytes are extended to int32 (not native int) on the
 evaluation stack.
+
+## Rules for IL Rewriters
+
+There are apis such as `System.Runtime.CompilerServices.RuntimeHelpers.CreateSpan<T>(...)` which require that the PE file have a particular structure. In particular, that api requires that the associated RVA of a FieldDef which is used to create a span must be naturally aligned over the data type that `CreateSpan` is instantiated over. There are 2 major concerns.
+
+1. That the RVA be aligned when the PE file is constructed. This may be achieved by whatever means is most convenient for the compiler.
+2. That in the presence of IL rewriters that the RVA remains aligned. This section descibes metadata which will be processed by IL rewriters in order to maintain the required alignment.
+
+In order to maintain alignment, if the field needs alignment to be preserved, the field must be of a type locally defined within the module which has a Pack (§II.10.7) value of the desired alignment. Unlike other uses of the .pack directive, in this circumstance the .pack specifies a minimum alignment.
```

Comment:
Should we enforce in the runtime that the field passed to `CreateSpan` is always of a locally defined type and the padding matches the alignment (unless ref emitted, I guess)?

That way we won't have IL Linker/crossgen2 breaking code that works in the absence of those tools. There's still a slight chance that the alignment in metadata is wrong, but it eliminates some of the risk. E.g. Roslyn sometimes avoids introducing the PrivateImplementationDetails type if it can use a primitive type and who knows if that codepath could kick in.